storage/src/vespa/storage/persistence/asynchandler.cpp

```diff
@@ -378,4 +397,38 @@ AsyncHandler::checkProviderBucketInfoMatches(const spi::Bucket& bucket, const ap
     return true;
 }
 
+MessageTracker::UP
+AsyncHandler::handleRemoveLocation(api::RemoveLocationCommand& cmd, MessageTracker::UP tracker) const
+{
+    tracker->setMetric(_env._metrics.removeLocation);
+
+    LOG(debug, "RemoveLocation(%s): using selection '%s'",
+        cmd.getBucketId().toString().c_str(),
+        cmd.getDocumentSelection().c_str());
+
+    spi::Bucket bucket(cmd.getBucket());
+    UnrevertableRemoveEntryProcessor::DocumentIdsAndTimeStamps to_remove;
+    UnrevertableRemoveEntryProcessor processor(to_remove);
+    BucketProcessor::iterateAll(_spi, bucket, cmd.getDocumentSelection(),
+                                std::make_shared<document::DocIdOnly>(),
+                                processor, spi::NEWEST_DOCUMENT_ONLY,tracker->context());
+
+    std::vector<std::future<std::unique_ptr<spi::Result>>> results;
+    results.reserve(to_remove.size());
+    for (auto & entry : to_remove) {
+        auto catcher = std::make_unique<spi::CatchResult>();
+        results.push_back(catcher->future_result());
+        _spi.removeAsync(bucket, entry.first, entry.second, tracker->context(), std::move(catcher));
+    }
+    for (auto & future : results) {
```

Comment:
Do we want to do a `results.back().wait()` before this (iff `results` is non-empty of course) to effectively guarantee we only put the calling thread to sleep once (assuming removes have their completion callbacks invoked in FIFO order)? This also ensures we don't throw any exceptions below before all pending ops to the bucket have completed.

Reply:
I do not think fifo order is guaranteed.
Anyway this is shortlived as I have another PR in the making that moves the asyncness further down.